### PR TITLE
feat: `ShouldAsk` for `Echo` and `Acknowledge`

### LINKED ISSUE #356: Support `should_ask` in `Echo` interaction
Feature request: conditionally display `Echo` interactions similar to how `BasicQuestion`, `Confirm`, etc. work. This would allow us to display text conditionally based on responses to previous questions.

> Confirm you like dogs? [y/N]
> <kbd>y</kbd>
> Great, now that we know you like dogs, here are some things you'll need to know:
> * Unlike cats, dogs actually like you
> * Unlike cats, dogs don't plot ways to kill you in your sleep
> 
> What breed of dog do you prefer? [English Setter]

### PR DESCRIPTION
**Pull request by mklaber**
* Create an abstract class from which `Echo` and `Acknowledge` inherit
* Add `should_ask` optional parameters
* Only display a message if `should_ask` is `None` or `True`

Closes #356

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,12 +4,13 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-## [Unreleased]
+## [0.12.0] - 2022-08-27
```
Comment (mklaber) [inline CHANGELOG.md:7]: @plannigan if you're comfortable with me releasing these 3.10 changes, I'll also try running the version bump process. Meanwhile, could you review the rest of the work?

File: docs/index.md [hunk 2]
```diff
@@ -18,6 +18,7 @@
     * As part of the text of a question
     * As part of the text of a default value
     * To decide if a question should be skipped
+    * To decide if a message should be displayed
```
Comment (plannigan) [inline docs/index.md:21]: Minor: I think it would be cleaner if was included as part of the previous item instead of adding a new one.

File: CHANGELOG.md (hunk 1 above)
Comment (plannigan) [reply CHANGELOG.md:7]: Can you please remove this line from the PR? The [repository has a process](https://wayfair-incubator.github.io/columbo/0.11.0/development-guide/#publishing-a-new-version) for updating the version number, which is incorporated in the the release process. Also, this is not the only place that the version number appears and needs to be incremented.

I'd also like to get #357 addressed in this release as well.

File: columbo/_interaction.py [hunk 3]
```diff
@@ -556,7 +619,8 @@ def get_answers(
 
     for interaction in interactions:
         if isinstance(interaction, (Echo, Acknowledge)):
-            interaction.display(result)
+            if interaction.should_ask(result):
```
Comment (plannigan) [inline columbo/_interaction.py:622]: Since, both cases now call `should_ask()`, that can be lifted out and called once before doing the `isinstance()` checks.

File: columbo/_interaction.py [hunk 4]
```diff
@@ -39,60 +39,121 @@ def _or_default(value, default: T) -> T:
     return default if isinstance(value, _Sentinel) else value
 
 
-class Echo:
-    """Display a message to the user."""
+class Displayable(ABC):
```
Comment (plannigan) [inline columbo/_interaction.py:42]: The base class for interactions that prompt the user is [exposed as part of the API](https://github.com/wayfair-incubator/columbo/blob/74265590f0bb57f6fd8f0c1b754d71e1d881fc8c/columbo/__init__.py#L15). Can you do that as well for `Displayable`.

File: columbo/_interaction.py [hunk 5]
```diff
@@ -39,60 +39,121 @@ def _or_default(value, default: T) -> T:
...
+        Should the user be displayed this message.
+
+        :param answers: The answers that have been provided this far.
+        :return: `True` if this message should be displayed
+        """
+        if self._should_ask is None:
+            return True
+        if callable(self._should_ask):
+            return self._should_ask(answers)
+        raise ValueError(f"Invalid value for should_ask: {self._should_ask}")
```
Comment (plannigan) [inline columbo/_interaction.py:76]: Minor: Instead of duplicating this functionality, can you extract it out into a private function used by `Displayable` & `Question`?

File: CHANGELOG.md (hunk 1 above)
Comment (mklaber) [reply CHANGELOG.md:7]: Ok. I'll also update CONTRIBUTING.md to make that point more clear.

File: docs/usage-guide/optional-questions-and-branching.md [hunk 6]
```diff
@@ -16,7 +19,8 @@ adventure. These choices introduce diverging paths of interactions that may or m
 ### Optional Questions
 
 The following is a basic example that has two optional questions that are not asked based on the answer to the first
-question.
+question. It also has an optional "echo" that is only displayed
```
[the remark this answers is not in the document]
Comment (mklaber) [reply docs/usage-guide/optional-questions-and-branching.md:22]: I wasn't referring to a class by name just as "two optional questions" wasn't referring to `BasicQuestion`s. I used quotation marks because it is otherwise awkward to use _echo_ as a noun here. I'll leave it as-is unless you have strong feelings (which it seems you don't).

File: columbo/_interaction.py (hunk 3 above)
Comment (mklaber) [reply columbo/_interaction.py:622]: I can't do that without failing a test:

```
FAILED tests/interaction_test.py::test_get_answers__unknown_interaction_provided - AttributeError: 'int' object has no attribute 'should_ask'
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -6,8 +6,8 @@ Please note we have a [code of conduct](CODE_OF_CONDUCT.md), please follow it in
 
 ## Pull Request Process
 
-1. Ensure the CI pipeline and all checks are passing.
-2. Update documentation with details of changes.
-3. Add an entry to the [change log](CHANGELOG.md).
+1. Ensure the CI pipeline and all checks are passing
+2. Update documentation with details of changes
+3. Add an entry to the [change log](CHANGELOG.md) using `[Unreleased]` as the version header
```
Comment (plannigan) [inline CONTRIBUTING.md:11]: Since this isn't directly related to this PR, can you do this as a different PR?

File: docs/usage-guide/optional-questions-and-branching.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (plannigan) [reply docs/usage-guide/optional-questions-and-branching.md:22]: Hmm. Maybe something "It also has an optional message that is only...". That phrasing could remove the confusion I had.

File: columbo/_interaction.py (hunk 3 above)
Comment (plannigan) [reply columbo/_interaction.py:622]: Oh right, I forgot about that. It was a minor thing, so we can leave it as is.

**Review by plannigan — APPROVED** [review]
Looks good, just one minor tweak to changelog.

File: CHANGELOG.md [hunk 8]
```diff
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Python version `3.10` tested during CI
 - Python version `3.10` added to package classifiers
+- `should_ask` kwarg for `Echo` and `Acknowledge` interactions ([#356](https://github.com/wayfair-incubator/columbo/issues/356))
```
[the remark this answers is not in the document]
Comment (mklaber) [reply CHANGELOG.md:13]: [copied from the best](https://github.com/wayfair-incubator/columbo/blob/a532829fb866046c108f1989fbedf084a2fd8658/CHANGELOG.md?plain=1#L30)

[the remark this answers is not in the document]
Comment (plannigan) [reply CHANGELOG.md]: 🤦 I guess I wasn't paying that much attention before. I'll put it on my list to clean up.
